compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeDefBuilderHelper.java

```diff
@@ -299,4 +307,72 @@ public static String getPackageAlias(SymbolEnv env, String compUnitName, Package
 
         return ""; // current module
     }
+
+    public static void populateStructureFields(Types types, SymbolTable symTable,
+                                                BLangAnonymousModelHelper anonymousModelHelper, Names names,
+                                                BLangStructureTypeNode structureTypeNode,
+                                                BStructureType structureType,
+                                                BStructureType origStructureType, Location pos,
+                                                SymbolEnv env, PackageID pkgID, Set<BType> unresolvedTypes,
+                                                long flag, boolean isImmutable, boolean isConstant) {
+        BTypeSymbol immutableStructureSymbol = structureType.tsymbol;
+        LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+        for (BField origField : origStructureType.fields.values()) {
+            BType fieldType;
+            if (isImmutable) {
+                fieldType = ImmutableTypeCloner.getImmutableType(pos, types, origField.type, env,
+                        env.enclPkg.packageID, env.scope.owner, symTable, anonymousModelHelper, names, unresolvedTypes);
```

Comment:
Shouldn't this happen for the const case also?

Reply:
Yeah. It happens when we are calling `getImmutableIntersectionType`.

Reply:
My point was if we even need the check at L321. But in the const case, this gets called for the mutable record, right? So I guess we still need this.